### PR TITLE
feat: AI vision sensor support

### LINKED ISSUE #46: AI Vision Support
## What's the motivation for this feature?
`vex-sdk` currently supports the AI Vision SDK functions. We should probably look into supporting that.

## Describe the solution you'd like
Use `vex-sdk`'s syscalls to implement an `AiVisionSensor` struct that abstracts over the smart device. This API might need to end up being a little different than the current `VisionSensor` API (which is far more scuffed internally).

## Describe the drawbacks, if any
- Barely anyone has access to this hardware at this point in time, making it difficult to test. Some features of it may also change in the future.
- We might want to share some functionality between the old vision sensor and this struct, depending on how similar the APIs are (I don't recall them being that similar, though). We should probably figure out how to best do this (should these sensors be in a single `vision` module?).

### PR DESCRIPTION
**Pull request by Gavin-Niederman**
## Describe the changes this PR makes. Why should it be merged?
This PR adds support for the AI vision sensor.
## Additional Context

Closes #46

### CODE REVIEW

**Comment by alexDickhans** [comment]
If you need testing I can test on the AI vision sensor our school has

**Comment by alexDickhans** [comment]
I have the vision sensor now if this is ready for testing

File: examples/ai_vision.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#![no_main]
+#![no_std]
+
+use core::time::Duration;
+
+use vexide::prelude::*;
+use vexide_devices::smart::ai_vision::{AiVisionColor, AiVisionSensor};
```
Comment (zabackary) [inline examples/ai_vision.rs:7]: I think most of the examples use the `vexide` crate's re-exports, so line 7 should probably do that as well. (Also, it looks like `vexide_devices` isn't even a dependency of the example crates.)

**Review by zabackary — COMMENTED** [review]
Just a couple of things I spotted. I know this is a draft, but it's probably better that these are caught earlier than later.

File: packages/vexide-devices/src/smart/ai_vision.rs [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+        self.brightness = brightness;
+        unsafe { vexDeviceAiVisionSensorSet(self.device, brightness, self.contrast) }
+        Ok(())
+    }
+
+    /// Sets a color code used to detect groups of colors.
+    /// # Note
+    /// This function will return an error if the given ID is not in the range [1, 8].
+    pub fn set_color_code(&mut self, id: u8, code: AiVisionColorCode) -> Result<()> {
+        if !(1..=8).contains(&id) {
```
[the remark this answers is not in the document]
Comment (Gavin-Niederman) [reply packages/vexide-devices/src/smart/ai_vision.rs:430]: This is an ID for a color code, not a signature. There are 8 slots for color codes.

File: packages/vexide-devices/src/smart/ai_vision.rs [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+        unsafe {
+            vexDeviceAiVisionCodeSet(self.device, &mut code as *mut _);
+        }
+
+        Ok(())
+    }
+
+    /// Returns the color code set on the AI Vision sensor with the given ID if it exists.
+    pub fn color_code(&self, id: u8) -> Result<Option<AiVisionColorCode>> {
+        if !(1..=8).contains(&id) {
```
[the remark this answers is not in the document]
Comment (zabackary) [reply packages/vexide-devices/src/smart/ai_vision.rs:611]: Never mind.

**Review by Tropix126 — APPROVED** [review]
LGTM. Hardware tested as of yesterday.

**Pull request merged**